src/par_iter/map.rs

```diff
@@ -112,16 +148,34 @@ impl<'m, P, MAP_OP, RET> Producer for MapProducer<'m, P, MAP_OP>
     }
 }
 
-impl<'m, P, MAP_OP, RET> IntoIterator for MapProducer<'m, P, MAP_OP>
+impl<'m, P, MAP_OP> IntoIterator for MapProducer<'m, P, MAP_OP>
     where P: Producer,
-          MAP_OP: Fn(P::Item) -> RET + Sync,
-          RET: Send,
+          MAP_OP: MapOp<P::Item>,
 {
-    type Item = RET;
-    type IntoIter = iter::Map<P::IntoIter, &'m MAP_OP>;
+    type Item = MAP_OP::Output;
+    type IntoIter = MapIter<'m, P::IntoIter, MAP_OP>;
 
     fn into_iter(self) -> Self::IntoIter {
-        self.base.into_iter().map(self.map_op)
+        MapIter {
+            base: self.base.into_iter(),
+            map_op: self.map_op,
+        }
```

Comment:
I guess having to write our own version of `map` is the casualty here, huh? Seems like a good trade-off though.